core/client/hammer/hammer.go

```diff
@@ -76,125 +93,159 @@ func New(ctx context.Context, dial DialFunc, callOptions CallOptions,
 		timeout:     timeout,
 		ktCli:       ktCli,
 		appID:       fmt.Sprintf("hammer_%v", time.Now().Format("2006-01-02/15:04:05")),
-		config:      config,
+		domain:      domain,
 
 		signers:        []*tink.KeysetHandle{keyset},
 		authorizedKeys: authorizedKeys.Keyset(),
 	}, nil
 }
 
 // Run runs operationCount with up to maxWorkers over ramp duration.
-func (h *Hammer) Run(ctx context.Context, operationCount, maxWorkers int, ramp time.Duration) {
-	jobs := gen(ctx, operationCount)
-	outputs := h.launchWorkers(ctx, jobs, maxWorkers, ramp)
-	h.printStats(1*time.Second, outputs)
-}
-
-// launchWorkers gradually adds workers to the worker pool over ramp up to workers.
-func (h *Hammer) launchWorkers(ctx context.Context, jobs <-chan int, workers int, ramp time.Duration) <-chan error {
-	outputs := make(chan error)
-
-	go func() {
-		rampTicker := time.NewTicker(ramp / time.Duration(workers))
-		var wg sync.WaitGroup
-		for h.workers < workers {
-			h.workers++
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				h.worker(ctx, jobs, outputs)
-			}()
-			<-rampTicker.C
-		}
-		rampTicker.Stop()
+// workers should roughly be (goal QPS / timeout seconds)
+func (h *Hammer) Run(ctx context.Context, numWorkers int, c Config) error {
+	workers, err := h.newWorkers(numWorkers)
+	if err != nil {
+		return err
+	}
 
-		wg.Wait()
-		close(outputs)
-	}()
+	// Batch Write users
+	log.Printf("Batch Write")
+	requests := genRequests(ctx, c.BatchWriteQPS, c.BatchWriteSize, c.BatchWriteCount, c.Duration)
+	handlers := make([]ReqHandler, 0, len(workers))
+	for _, w := range workers {
+		handlers = append(handlers, w.writeOp)
+	}
+	generateReport(ctx, requests, handlers)
+
+	// Write users
+	log.Printf("User Write")
+	requests = genRequests(ctx, c.WriteQPS, 1, c.WriteCount, c.Duration)
+	handlers = make([]ReqHandler, 0, len(workers))
+	for _, w := range workers {
+		handlers = append(handlers, w.writeOp)
+	}
+	generateReport(ctx, requests, handlers)
+
+	// Read users
+	log.Printf("User Read")
+	requests = genRequests(ctx, c.ReadQPS, 1, c.ReadCount, c.Duration)
+	handlers = make([]ReqHandler, 0, len(workers))
+	for _, w := range workers {
+		handlers = append(handlers, w.readOp)
+	}
+	generateReport(ctx, requests, handlers)
+
+	// History
+	log.Printf("User Audit History")
+	requests = genRequests(ctx, c.HistoryQPS, c.HistoryBatchSize, c.HistoryCount, c.Duration)
+	handlers = make([]ReqHandler, 0, len(workers))
+	for _, w := range workers {
+		handlers = append(handlers, w.historyOp)
+	}
+	generateReport(ctx, requests, handlers)
 
-	return outputs
+	return nil
 }
 
-// gen produces a channel of integers from 0 up to count-1 and then closes.
-func gen(ctx context.Context, count int) <-chan int {
-	jobs := make(chan int)
+func genRequests(ctx context.Context, qps, batch, count int, duration time.Duration) <-chan request {
```

Comment:
Maybe `requestGenerator` rather than `genRequests` to give a hint that it hasn't generated all of the requests by the time it returns.

Reply:
done

Reply:
done